src/main/ruby/truffleruby/core/array.rb

```diff
@@ -1007,67 +1007,93 @@ def sample(count=undefined, options=undefined)
   end
 
   def sample_many(count, rng)
-    if size / count > 3
-      abandon = false
+    swap = false
+    quad = false # three implementations
 
+    if count <= 70 # choice determined experimentally
+      if 2.0 * size / count  <= count  + 13
+        swap = true
+      else
+        quad = true
+      end
+    else
+      if size <= -1100.0 + 59.5 * count
+        swap = true
+      end
+    end
+
+    if swap
+      result = Array.new(self)
+
+      count.times do |c|
+        result.__send__ :swap, c, rng.rand(size)
+      end
+
+      count == size ? result : result[0, count]
+    else
       result = Array.new count
       i = 1
 
       result[0] = rng.rand(size)
-      while i < count
-        k = rng.rand(size)
 
-        spin = false
-        spin_count = 0
+      if quad #quadratic time due to linear time collision check but low overhead
+        while i < count
+          k = rng.rand(size)
 
-        while true # rubocop:disable Lint/LiteralAsCondition
-          j = 0
-          while j < i
-            if k == result[j]
-              spin = true
-              break
-            end
+          spin = false
 
-            j += 1
-          end
+          while true # rubocop:disable Lint/LiteralAsCondition
```

Comment:
(not for this PR, I'll do it)
I'm thinking we should just disable this rule globally, it seems annoying and very rarely useful.
There seem to be no way to configure it only for `if/unless` and not `while/until`.